### PR TITLE
Fix `setKernelMacServer` NetworkServiceServer chain element and add tests to cover

### PR DESCRIPTION
**Pull request by denis-tingajkin**
Signed-off-by: Denis Tingajkin <denis.tingajkin@xored.com>

The first part of splitting PR https://github.com/networkservicemesh/sdk-vppagent/pull/38

### CODE REVIEW

File: pkg/networkservice/connectioncontextkernel/ethernetcontext/setmacaddress/server.go [hunk 1]
```diff
@@ -14,11 +16,12 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
-// Package macaddress provides networkservice chain elements for setting the mac address on kernel interfaces
-package macaddress
+// Package setmacaddress provides networkservice chain elements for setting the mac address on kernel interfaces
+package setmacaddress
```
Comment (edwarnicke) [inline pkg/networkservice/connectioncontextkernel/ethernetcontext/setmacaddress/server.go]: Question, why did you rename this setmacaddress ?

File: pkg/networkservice/connectioncontextkernel/ethernetcontext/setmacaddress/server.go [hunk 2]
```diff
@@ -57,26 +60,32 @@ type setVppMacServer struct{}
...
-func (s *setVppMacServer) Close(ctx context.Context, conn *connection.Connection) (*empty.Empty, error) {
-	conf := vppagent.Config(ctx)
-	if mechanism := kernel.ToMechanism(conn.GetMechanism()); mechanism != nil && len(conf.GetLinuxConfig().GetInterfaces()) > 0 {
-		conf.GetLinuxConfig().GetInterfaces()[len(conf.GetVppConfig().GetInterfaces())-1].PhysAddress = conn.GetContext().GetEthernetContext().GetDstMac()
+func (s *setKernelMacServer) setDstMac(ctx context.Context, conn *connection.Connection) {
+	ethernetContext := conn.GetContext().GetEthernetContext()
+	config := vppagent.Config(ctx)
+	if ethernetContext != nil && ethernetContext.DstMac != "" {
+		interfaceName := fmt.Sprintf("DST-%v", conn.GetId())
+		for _, iface := range config.LinuxConfig.Interfaces {
```
Comment (edwarnicke) [inline pkg/networkservice/connectioncontextkernel/ethernetcontext/setmacaddress/server.go]: Why are we looping over all interfaces here?

Comment (denis-tingajkin) [reply pkg/networkservice/connectioncontextkernel/ethernetcontext/setmacaddress/server.go]: To exclude setting mac address for veth interfaces for example

File: pkg/networkservice/connectioncontextkernel/ethernetcontext/setmacaddress/server.go (hunk 1 above)
Comment (denis-tingajkin) [reply pkg/networkservice/connectioncontextkernel/ethernetcontext/setmacaddress/server.go]: > Question, why did you rename this setmacaddress ?

I did rename this for name convention.
The next step is to add `getmacaddress` package https://github.com/networkservicemesh/sdk-vppagent/pull/38#discussion_r370341434. I guess it should be consistent with `macaddress` package.

Comment (edwarnicke) [reply pkg/networkservice/connectioncontextkernel/ethernetcontext/setmacaddress/server.go]: Lets leave this as macaddress.  It follows the convention *other* places in the code so far.  getmac is sort of an odd outlier itself, as we don't do that for most things.

File: pkg/networkservice/connectioncontextkernel/ethernetcontext/setmacaddress/server.go (hunk 2 above)
Comment (edwarnicke) [reply pkg/networkservice/connectioncontextkernel/ethernetcontext/setmacaddress/server.go]: The problem is it relies on a naming convention, one that is already different from what is here:

https://github.com/edwarnicke/sdk-vppagent/blob/master/pkg/networkservice/mechanisms/kernel/kernelvethpair/server.go#L45

Rather than baking naming conventions in, the previous approach was simply this: the last Linux Interface is the one we apply config to.

File: pkg/networkservice/connectioncontextkernel/ethernetcontext/setmacaddress/server.go (hunk 1 above)
Comment (denis-tingajkin) [reply pkg/networkservice/connectioncontextkernel/ethernetcontext/setmacaddress/server.go]: By `getmac` I mean the chain element described in https://github.com/networkservicemesh/sdk-vppagent/issues/27.

As I understand this element should find created by vpp-agent mac address of interface and put it into ethernetcontext.DstMac. And it should be useful for forwarders implementation.
Did I miss something?

File: pkg/networkservice/connectioncontextkernel/ethernetcontext/setmacaddress/server.go (hunk 2 above)
Comment (denis-tingajkin) [reply pkg/networkservice/connectioncontextkernel/ethernetcontext/setmacaddress/server.go]: Do you suggest use index convention?

Comment (denis-tingajkin) [reply pkg/networkservice/connectioncontextkernel/ethernetcontext/setmacaddress/server.go]: Using the last interface looking is not clear to me. Is it some pattern?
Can we just put the current interface into context? For example

context.WithDestenationInterface(ctx, inteface)
...
interface := vppagent.DestenationInterface(ctx)
interface.Set(...)

Comment (denis-tingajkin) [reply pkg/networkservice/connectioncontextkernel/ethernetcontext/setmacaddress/server.go]: Finally got it

File: pkg/networkservice/connectioncontextkernel/ethernetcontext/macaddress/server.go [hunk 3]
```diff
@@ -57,26 +58,28 @@ type setVppMacServer struct{}
...
 }
 
-func (s *setVppMacServer) Close(ctx context.Context, conn *connection.Connection) (*empty.Empty, error) {
-	conf := vppagent.Config(ctx)
-	if mechanism := kernel.ToMechanism(conn.GetMechanism()); mechanism != nil && len(conf.GetLinuxConfig().GetInterfaces()) > 0 {
-		conf.GetLinuxConfig().GetInterfaces()[len(conf.GetVppConfig().GetInterfaces())-1].PhysAddress = conn.GetContext().GetEthernetContext().GetDstMac()
+func (s *setKernelMacServer) setDstMac(ctx context.Context, conn *connection.Connection) {
+	ethernetContext := conn.GetContext().GetEthernetContext()
+	config := vppagent.Config(ctx)
+	if ethernetContext != nil && ethernetContext.DstMac != "" {
```
Comment (edwarnicke) [inline pkg/networkservice/connectioncontextkernel/ethernetcontext/macaddress/server.go]: Why are we pulling this out into a function.  We don't have to do all this nil checking if we use the GRPC generated Get functions.  Just leave it inline.

File: pkg/networkservice/connectioncontextkernel/ethernetcontext/macaddress/server_test.go [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+	"github.com/networkservicemesh/sdk-vppagent/pkg/networkservice/vppagent"
+
+	"github.com/networkservicemesh/networkservicemesh/controlplane/api/connection"
+	"github.com/networkservicemesh/networkservicemesh/controlplane/api/connection/mechanisms/kernel"
+	"github.com/networkservicemesh/networkservicemesh/controlplane/api/connectioncontext"
+	"github.com/networkservicemesh/networkservicemesh/controlplane/api/networkservice"
+)
+
+func TestServerBasic(t *testing.T) {
+	ctx := vppagent.WithConfig(context.Background())
```
Comment (edwarnicke) [inline pkg/networkservice/connectioncontextkernel/ethernetcontext/macaddress/server_test.go]: Please use

https://github.com/networkservicemesh/sdk-vppagent/blob/master/pkg/networkservice/vppagent/server.go#L33

In a chain instead of exposing withConfig.

File: pkg/networkservice/connectioncontextkernel/ethernetcontext/macaddress/server.go [hunk 5]
```diff
@@ -57,26 +58,25 @@ type setVppMacServer struct{}
...
-		conf.GetLinuxConfig().GetInterfaces()[len(conf.GetVppConfig().GetInterfaces())-1].PhysAddress = conn.GetContext().GetEthernetContext().GetDstMac()
+func (s *setKernelMacServer) Close(ctx context.Context, conn *connection.Connection) (*empty.Empty, error) {
+	if mechanism := kernel.ToMechanism(conn.GetMechanism()); mechanism != nil {
+		s.setDstMac(ctx, conn)
 	}
-	return next.Client(ctx).Close(ctx, conn)
+	return next.Server(ctx).Close(ctx, conn)
+}
+
+func (s *setKernelMacServer) setDstMac(ctx context.Context, conn *connection.Connection) {
```
Comment (edwarnicke) [inline pkg/networkservice/connectioncontextkernel/ethernetcontext/macaddress/server.go]: Please leave this inline rather than pulling it out into a separate setDstMac function here.  It's small and simple and more comprehensible inline in Request/Close

File: pkg/networkservice/connectioncontextkernel/ethernetcontext/macaddress/server.go (hunk 3 above)
Comment (denis-tingajkin) [reply pkg/networkservice/connectioncontextkernel/ethernetcontext/macaddress/server.go]: thanks, fixed

File: pkg/networkservice/connectioncontextkernel/ethernetcontext/macaddress/server.go (hunk 5 above)
Comment (denis-tingajkin) [reply pkg/networkservice/connectioncontextkernel/ethernetcontext/macaddress/server.go]: fixed.

File: pkg/networkservice/connectioncontextkernel/ethernetcontext/macaddress/server_test.go (hunk 4 above)
Comment (denis-tingajkin) [reply pkg/networkservice/connectioncontextkernel/ethernetcontext/macaddress/server_test.go]: fixed.

File: pkg/networkservice/vppagent/client.go [hunk 6]
```diff
@@ -17,12 +19,26 @@
...
+	return next.Client(ctx).Request(withConfig(ctx), in, opts...)
+}
+
+func (c configClient) Close(ctx context.Context, in *connection.Connection, opts ...grpc.CallOption) (*empty.Empty, error) {
+	return next.Client(ctx).Close(withConfig(ctx), in, opts...)
+}
+
 // NewClient - inserts a vppagent *configurator.Config into the GRPC call context.Context
 func NewClient() networkservice.NetworkServiceClient {
-	return adapters.NewServerToClient(&configServer{})
```
Comment (denis-tingajkin) [inline pkg/networkservice/vppagent/client.go]: https://github.com/networkservicemesh/sdk/issues/61

Comment (edwarnicke) [reply pkg/networkservice/vppagent/client.go]: Nice catch!

File: pkg/networkservice/vppagent/context.go [hunk 7]
```diff
@@ -32,7 +32,7 @@ const (
 	configKey contextKeyType = "configKey"
 )
 
-// WithConfig gets vppagent config from context
+// withConfig gets vppagent config from context
```
Comment (edwarnicke) [inline pkg/networkservice/vppagent/context.go]: Thanks for correcting the comment here.

File: pkg/networkservice/connectioncontext/client.go [hunk 8]
```diff
@@ -17,9 +17,10 @@
 package connectioncontext
 
 import (
-	"github.com/networkservicemesh/networkservicemesh/controlplane/api/networkservice"
 	"github.com/networkservicemesh/sdk/pkg/networkservice/core/chain"
 
+	"github.com/networkservicemesh/networkservicemesh/controlplane/api/networkservice"
```
Comment (edwarnicke) [inline pkg/networkservice/connectioncontext/client.go]: I'm curious why these header orders changed here and in server.go.  Not going to hold this PR up over it, but it suggests perhaps the goimports linting isn't strict enough.

File: pkg/networkservice/connectioncontextkernel/ethernetcontext/macaddress/client.go [hunk 9]
```diff
@@ -60,23 +62,20 @@ func NewClient() networkservice.NetworkServiceClient {
...
-		index := len(conf.GetLinuxConfig().GetInterfaces()) - 1
-		conf.GetLinuxConfig().GetInterfaces()[index+1].PhysAddress = conn.GetContext().GetEthernetContext().GetSrcMac()
+func (c *setKernelMacClient) Request(ctx context.Context, request *networkservice.NetworkServiceRequest, opts ...grpc.CallOption) (*connection.Connection, error) {
+	config := vppagent.Config(ctx)
+	if mechanism := kernel.ToMechanism(request.GetConnection().GetMechanism()); mechanism != nil && len(config.GetLinuxConfig().GetInterfaces()) > 0 {
+		current := len(config.LinuxConfig.Interfaces) - 1
+		config.LinuxConfig.Interfaces[current].PhysAddress = request.GetConnection().GetContext().GetEthernetContext().GetSrcMac()
 	}
-	return conn, nil
+	return next.Client(ctx).Request(ctx, request, opts...)
```
Comment (edwarnicke) [inline pkg/networkservice/connectioncontextkernel/ethernetcontext/macaddress/client.go]: You really have to call the next.Client here *before* you setup the config because the fully completed connection will come *back* from the return to that call, it doesn't come *in* on the Request for the client.

File: pkg/networkservice/connectioncontext/client.go (hunk 8 above)
Comment (denis-tingajkin) [reply pkg/networkservice/connectioncontext/client.go]: Probably it is a problem of configuring `golangci-linter`.  I can take a look at this issue in separate PR if you don't mind.

Comment (edwarnicke) [reply pkg/networkservice/connectioncontext/client.go]: Definitely shouldn't hold up this PR.
